### PR TITLE
Chuck/web 1614 unify workspaceproject role creation

### PR DESCRIPTION
**Pull request by cdriesler**
## Description & motivation

- We have made decisions around how workspace roles will interact and imply project roles (for now) (before teams)

## Changes:

- Implements logic from [The Figma](https://www.figma.com/board/IjsE2t92YHmyTtoEqmXRsX/Workspace-Roles?t=LT7uU66QwiwRT1zD-0)

### CODE REVIEW

File: packages/server/modules/workspaces/services/management.ts [hunk 1]
```diff
@@ -19,8 +19,8 @@ import { MaybeNullOrUndefined, Roles } from '@speckle/shared'
 import cryptoRandomString from 'crypto-random-string'
 import {
   deleteStream,
-  grantStreamPermissions as repoGrantStreamPermissions,
-  revokeStreamPermissions as repoRevokeStreamPermissions
+  grantStreamPermissionsFactory,
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/services/management.ts:22]: we should not invite repositories into services directly, this should be a domain / operation interface type import

File: packages/server/modules/workspaces/services/management.ts [hunk 2]
```diff
@@ -323,12 +325,12 @@ export const updateWorkspaceRoleFactory =
...
+      nextRole !== Roles.Workspace.Admin
     ) {
       throw new WorkspaceAdminRequiredError()
     }
 
-    if (role !== Roles.Workspace.Guest) {
+    if (nextRole !== Roles.Workspace.Guest) {
       const workspace = await getWorkspaceWithDomains({ id: workspaceId })
       const verifiedDomains = workspace?.domains.filter((domain) => domain?.verified)
       if (
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/services/management.ts:336]: should be replaced with `userEmailsCompliantWithWorkspaceDomains` from the soon to be merged #2797 PR

File: packages/server/modules/workspaces/services/management.ts [hunk 3]
```diff
@@ -349,36 +351,121 @@ export const updateWorkspaceRoleFactory =
...
-    const isFirstWorkspaceRole = !workspaceRoles.some((role) => role.userId === userId)
+    // Update project roles
+    const currentRole = workspaceRoles.find((acl) => acl.userId === userId)?.role
 
-    if (!isFirstWorkspaceRole || role === Roles.Workspace.Guest) {
-      // Guests do not get roles for existing workspace projects
-      return
-    }
+    // TODO: This is a workspace setting
+    const defaultProjectRole = Roles.Stream.Contributor
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/services/management.ts:366]: lets lower this to stream reviewer

File: packages/server/modules/workspaces/services/management.ts [hunk 4]
```diff
@@ -349,36 +351,121 @@ export const updateWorkspaceRoleFactory =
...
 
     const queryAllWorkspaceProjectsGenerator = queryAllWorkspaceProjectsFactory({
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/services/management.ts:368]: this must be an injected dependency we rely on

**Review by gjedlicska — CHANGES_REQUESTED** [review]

**Comment by cdriesler** [comment]
Ready for final thumbs up @gjedlicska 🫡

**Pull request merged**